source/jormungandr/jormungandr/scenarios/experimental.py

```diff
@@ -118,18 +101,103 @@ def update_crowfly_duration(instance, fallback_list, mode, stop_area_uri):
+            continue
+        previous_section_begin = j.arrival_date_time
+        for s in j.sections:
+            from copy import deepcopy
+            o = deepcopy(s.origin)
+            d = deepcopy(s.destination)
+            s.origin.CopyFrom(d)
+            s.destination.CopyFrom(o)
+            s.end_date_time = previous_section_begin
+            previous_section_begin = s.begin_date_time = s.end_date_time - s.duration
+    return res
+
 
 class Scenario(new_default.Scenario):
 
     def __init__(self):
         super(Scenario, self).__init__()
 
-    def _get_direct_path(self, instance, mode, pt_object_origin,
-                         pt_object_destination, datetime, clockwise):
+    def _get_direct_path(self, instance, mode, pt_object_origin, pt_object_destination, datetime, clockwise):
         # TODO: cache by (mode, origin, destination) and redate with datetime and clockwise
         return instance.street_network_service.direct_path(mode, pt_object_origin,
                                                            pt_object_destination, datetime, clockwise)
 
+    def _build_journey(self, journey, instance, _from, to, dep_mode, arr_mode):
+        import copy
+        origins = g.origins_fallback[dep_mode]
+        destinations = g.destinations_fallback[arr_mode]
+
+        departure = journey.sections[0].origin
+        arrival = journey.sections[-1].destination
+        last_section_end = journey.sections[-1].end_date_time
+
+        journey.duration = journey.duration + origins[departure.uri] + destinations[arrival.uri]
+        journey.departure_date_time = journey.departure_date_time - origins[departure.uri]
+        journey.arrival_date_time = journey.arrival_date_time + destinations[arrival.uri]
+
+        departure_dp_key = (dep_mode, _from.uri, departure.uri, journey.departure_date_time, True)
+
+        if _from.uri != departure.uri:
+            if origins[departure.uri] == 0:
+                journey.sections.extend([create_crowfly(_from, departure, journey.departure_date_time,
+                                         journey.sections[0].begin_date_time)])
+            elif not g.fallback_direct_path.get(departure_dp_key):
+                g.fallback_direct_path[departure_dp_key] = self._get_direct_path(instance,
+                                                                                 dep_mode,
+                                                                                 _from,
+                                                                                 departure,
+                                                                                 journey.departure_date_time,
+                                                                                 True)
+            if origins[departure.uri] != 0:
+                departure_direct_path = copy.deepcopy(g.fallback_direct_path[departure_dp_key])
+                _extend_pt_sections_with_direct_path(journey, departure_direct_path)
```

Comment:
- [ ] move the cache `into _get_direct_path`

- [ ] handle the crowfly in particular cases